File: src/main/java/net/maple/handlers/misc/GroupMessageHandler.java
package net.maple.handlers.misc;

import client.Character;
import client.Client;
import lombok.RequiredArgsConstructor;
import net.maple.SendOpcode;
import net.maple.handlers.PacketHandler;
import util.packet.Packet;
import util.packet.PacketReader;
import util.packet.PacketWriter;

public class GroupMessageHandler extends PacketHandler {

    @RequiredArgsConstructor
    private static final class ChatGroupType {
        public static final byte FRIEND = 0x00;
        public static final byte PARTY = 0x01;
        public static final byte GUILD = 0x02;
        public static final byte ALLIANCE = 0x03;
        public static final byte COUPLE = 0x04;
        public static final byte TO_COUPLE = 0x05;
        public static final byte EXPEDITION = 0x06;
    }

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        Character chr = c.getCharacter();
        reader.readInteger(); // timestamp
        byte type = reader.readByte(); // chat type
        int size = reader.readByte(); // count members
        for (int i = 0; i < size; i++) {
            reader.readInteger(); // cid
        }
        String message = reader.readMapleString(); // the message

        switch (type) {
            case ChatGroupType.FRIEND:
                chr.getFriendList().sendMessage(multiChat(ChatGroupType.FRIEND, chr.getName(), message));
                break;
            case ChatGroupType.PARTY:
                if (chr.getParty() != null) {
                    chr.getParty().sendMessage(multiChat(ChatGroupType.PARTY, chr.getName(), message), chr.getId());
                }
            case ChatGroupType.GUILD:
                if (chr.getGuild() != null) {
                    chr.getGuild().broadcast(multiChat(ChatGroupType.GUILD, chr.getName(), message), chr);
                }
                break;
            default:
                System.out.println("Unknown GroupMessageHandler type (" + type + ") from " + chr.getName());
                break;

        }
    }

    public static Packet multiChat(byte type, String name, String message) {
        PacketWriter pw = new PacketWriter(8);

        pw.writeHeader(SendOpcode.GROUP_MESSAGE);
        pw.write(type);
        pw.writeMapleString(name);
        pw.writeMapleString(message);

        return pw.createPacket();
    }
}

File: src/main/java/net/maple/handlers/group/GroupMessageHandler.kt
package net.maple.handlers.group

import client.Client
import net.maple.SendOpcode
import net.maple.handlers.PacketHandler
import util.packet.Packet
import util.packet.PacketReader
import util.packet.PacketWriter

class GroupMessageHandler : PacketHandler {

    private object ChatGroupType {
        const val FRIEND: Byte = 0x00
        const val PARTY: Byte = 0x01
        const val GUILD: Byte = 0x02
        const val ALLIANCE: Byte = 0x03
        const val COUPLE: Byte = 0x04
        const val TO_COUPLE: Byte = 0x05
        const val EXPEDITION: Byte = 0x06
    }

    override fun handlePacket(reader: PacketReader, c: Client) {
        val chr = c.character
        reader.readInteger() // timestamp
        val type = reader.readByte() // chat type
        val size = reader.readByte().toInt() // count members
        for (i in 0 until size) {
            reader.readInteger() // cid
        }
        val message = reader.readMapleString() // the message

        when (type) {
            ChatGroupType.FRIEND -> {
                chr.friendList.sendMessage(multiChat(ChatGroupType.FRIEND, chr.name, message))
            }
            ChatGroupType.PARTY -> {
                val party = chr.party ?: return
                party.sendMessage(multiChat(ChatGroupType.PARTY, chr.name, message), chr.id)
            }
            ChatGroupType.GUILD -> {
                val guild = chr.guild ?: return
                guild.broadcast(multiChat(ChatGroupType.GUILD, chr.name, message), chr)
            }
            else -> println("Unknown GroupMessageHandler type ($type) from ${chr.name}")
        }
    }

    companion object {
        fun multiChat(type: Byte, name: String, message: String): Packet {
            val pw = PacketWriter(8)

            pw.writeHeader(SendOpcode.GROUP_MESSAGE)
            pw.write(type.toInt())
            pw.writeMapleString(name)
            pw.writeMapleString(message)

            return pw.createPacket()
        }
    }
}